src/main/java/org/spongepowered/common/resource/pack/SpongeFileSystemPack.java

```diff
@@ -0,0 +1,151 @@
+    }
+
+    /**
+     * Creates a new filesystem from the root path lazily.
+     *
+     * @return The filesystem
+     * @throws IOException If the filesystem could not be created
+     */
+    private FileSystem getFileSystem() throws IOException {
+        if (fileSystem == null) {
+            if (Files.isRegularFile(rootPath)) {
+                fileSystem = FileSystems.newFileSystem(URI.create("jar:" + rootPath.toUri()), Collections.emptyMap());
+            } else if (Files.isDirectory(rootPath)) {
+                fileSystem = FileSystems.newFileSystem(rootPath.toUri(), Collections.emptyMap());
+            } else {
+                throw new IOException("Unsupported filesystem: " + rootPath);
+            }
+        }
+        return fileSystem;
+    }
+
+    private Path getPackRoot(PackType type) throws IOException {
+        return getFileSystem().getPath(((ResourcePackType) (Object) type).getDirectoryName());
+    }
+
+    private Path getFilePath(String path) throws IOException {
+        return getFileSystem().getPath(path);
+    }
+
+    @Override
+    protected InputStream openStream(String path) throws IOException {
+        return Files.newInputStream(getFilePath(path));
+    }
+
+    @Override
+    protected boolean exists(String path) {
+        try {
+            return Files.exists(getFilePath(path));
+        } catch (IOException e) {
+            return false;
+        }
+    }
+
+    @Override
+    public Set<String> getNamespaces(PackType type) {
+        try {
+            return Files.list(getPackRoot(type))
+                    .map(Path::getFileName)
+                    .map(Object::toString)
+                    .filter(s -> {
+                        if (s.equals(s.toLowerCase(Locale.ROOT))) {
+                            return true;
+                        } else {
+                            SpongeCommon.getLogger().warn("Pack: ignored non-lowercased namespace: {} in {}", s, this.rootPath);
+                            return false;
+                        }
+                    })
+                    .collect(Collectors.toSet());
+        } catch (IOException e) {
+            return Collections.emptySet();
+        }
+    }
+
+    @Override
+    public void close() throws IOException {
+        if (fileSystem != null) {
+            fileSystem.close();
+        }
+    }
+
+    @Override
+    public Collection<ResourcePath> find(PackType type, String prefix, int depth, Predicate<String> filter) {
+        try {
+            Path root = getPackRoot(type);
+            return getNamespaces(type).stream()
+                    .flatMap(namespace -> {
+                        try {
+                            return Files.walk(root.resolve(namespace).resolve(prefix), depth)
+                                    .filter(s -> !s.getFileName().toString().endsWith(".mcmeta"))
+                                    // todo is the path relative to start?
+                                    .map(Object::toString)
```

Comment:
I need to do some testing on this to see if the path is relative. If it is not, I will have to relativize it manually before converting it to a string.